docs/codeql/codeql-language-guides/customizing-library-models-for-csharp.rst

```diff
@@ -0,0 +1,331 @@
+       ...
+   }
+
+We need to add a tuple to the ``sourceModel``\(namespace, type, subtypes, name, signature, ext, output, kind, provenance) extensible predicate by updating a data extension file.
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/csharp-all
+         extensible: sourceModel
+       data:
+         - ["System.Net.Sockets", "TcpClient", False, "GetStream", "()", "", "ReturnValue", "remote", "manual"]
+
+
+Since we are adding a new source, we need to add a tuple to the ``sourceModel`` extensible predicate.
+The first five values identify the callable (in this case a method) to be modeled as a source.
+
+- The first value ``System.Net.Sockets`` is the namespace name.
+- The second value ``TcpClient`` is the name of the class (type) that contains the source.
+- The third value ``False`` is a flag that indicates whether or not the source also applies to all overrides of the method.
+- The fourth value ``GetStream`` is the method name.
+- The fifth value ``()`` is the method input type signature.
+
+The sixth value should be left empty and is out of scope for this documentation.
+The remaining values are used to define the ``access path``, the ``kind``, and the ``provenance`` (origin) of the source.
+
+- The seventh value ``ReturnValue`` is the access path to the return of the method, which means that it is the return value that should be considered a source of tainted input.
+- The eighth value ``remote`` is the kind of the source. The source kind is used to define the threat model where the source is in scope. ``remote`` applies to many of the security related queries as it means a remote source of untrusted data. As an example the SQL injection query uses ``remote`` sources.
+- The ninth value ``manual`` is the provenance of the source, which is used to identify the origin of the source.
+
+Example: Add flow through the ``Concat`` method
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+This example shows how the C# query pack models flow through a method for a simple case.
+This pattern covers many of the cases where we need to summarize flow through a method that is stored in a library or framework outside the repository.
+
+.. code-block:: csharp
+
+   public static void TaintFlow(string s1, string s2) {
+       string t = String.Concat(s1, s2); // There is taint flow from s1 and s2 to t.
+       ...
+   }
+
+We need to add tuples to the ``summaryModel``\(namespace, type, subtypes, name, signature, ext, input, output, kind, provenance) extensible predicate by updating a data extension file:
+
+.. code-block:: yaml
+
+   extensions:
+     - addsTo:
+         pack: codeql/csharp-all
+         extensible: summaryModel
+       data:
+         - ["System", "String", False, "Concat", "(System.Object,System.Object)", "", "Argument[0]", "ReturnValue", "taint", "manual"]
+         - ["System", "String", False, "Concat", "(System.Object,System.Object)", "", "Argument[1]", "ReturnValue", "taint", "manual"]
```

Comment:
It should be possible two write a single combined summary:
```
- ["System", "String", False, "Concat", "(System.Object,System.Object)", "", "Argument[0,1]", "ReturnValue", "taint", "manual"]
```

Reply:
I have changed the model in the example to use `Argument[0,1]`, as well as updated a note to reference the shorthand syntax.

See: [a474b09](https://github.com/github/codeql/pull/15563/commits/a474b093b817d34cc97dd7a66d19b99024bd2659)